Python/fileutils.c

```diff
@@ -2450,6 +2470,47 @@ _Py_set_blocking(int fd, int blocking)
     return -1;
 }
 #else   /* MS_WINDOWS */
+int
+_Py_get_blocking(int fd)
+{
+    HANDLE handle;
+    DWORD mode;
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE) {
+        return -1;
+    }
+
+    if (!GetNamedPipeHandleStateW(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    return mode & PIPE_WAIT;
```

Comment:
The GIL has to be released here. Also, I messed up the return value. It has to test for `PIPE_NOWAIT` (1) and invert the result.

```suggestion
    DWORD mode;
    BOOL success;

    handle = _Py_get_osfhandle(fd);
    if (handle == INVALID_HANDLE_VALUE) {
        return -1;
    }

    Py_BEGIN_ALLOW_THREADS
    success = GetNamedPipeHandleStateW(handle, &mode,
                                       NULL, NULL, NULL, NULL, 0);
    Py_END_ALLOW_THREADS
    
    if (!success) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
    
    return !(mode & PIPE_NOWAIT);
```

The GIL has to be release because get/set pipe operations use the same synchronization lock that's used by synchronous reads and writes. Thus these operations can block if another thread is waiting for I/O. For example, using the code from this pull request:

```pycon
>>> import os, threading
>>> r, w = os.pipe()
>>> threading.Thread(target=os.read, args=(r, 1)).start()
>>> os.get_blocking(r) # This blocks.
```